arch/x86_64/inject.c

```diff
@@ -0,0 +1,205 @@
+#include <string.h>
+#include <unistd.h>
+#include <sys/user.h>
+#include <wait.h>
+#include <sys/time.h>
+#include "utils/inject.h"
+#include "utils/ptrace.h"
+#include "utils/utils.h"
+
+#define SO_LIBC_NAME "libc.so.6"
+#define SO_LIBDL_NAME "libdl.so.2"
+
+extern void inject_content(void);
+extern const void inject_content_end;
+
+int inject(char *libname, pid_t pid)
+{
+	struct user_regs_struct oldregs, regs;
+	struct user_regs_struct malloc_regs;
+	struct user_regs_struct dlopen_regs;
+	int lib_path_length, mypid;
+	long my_libc_addr, my_libdl_addr;
+	long malloc_addr, free_addr, dlopen_addr;
+	long malloc_offset, free_offset, dlopen_offset;
+	long target_libc_addr, target_libdl_addr; 
+	long target_malloc_addr, target_free_addr;
+	long target_dlopen_addr, addr;
+	unsigned long long target_buf, lib_addr;
+	char *newcode, *backup;
+	char *lib_path = realpath(libname, NULL);
+	pid_t target = 0;
+
+	// figure out the size of contents to be injected.
+	// so, we know how big of a buffer to allocate.
+	uint32_t inject_content_size = (uintptr_t)&inject_content_end - (uintptr_t)inject_content;
+	size_t inject_shared_lib_size = inject_content_size;
+
+	if (!lib_path)
+		pr_err_ns("can't find file %s", libname);
+
+	target = pid;
+
+	lib_path_length = strlen(lib_path) + 1;
+	mypid = getpid();
+	my_libc_addr = get_libc_addr(mypid);
+	my_libdl_addr = get_libdl_addr(mypid);
+
+	/*
+	 * find the addresses of the syscalls that we'd like to use
+	 * inside the target, as loaded inside THIS process
+	 * (i.e. NOT the target process)
+	 */
+	malloc_addr = get_function_addr(SO_LIBC_NAME, "malloc");
+	free_addr = get_function_addr(SO_LIBC_NAME, "free");
+	dlopen_addr = get_function_addr(SO_LIBDL_NAME, "dlopen");
+
+	// use the base address of libc to calculate offsets for
+	// the syscalls we want to use
+	malloc_offset = malloc_addr - my_libc_addr;
+	free_offset = free_addr - my_libc_addr;
+	dlopen_offset = dlopen_addr - my_libdl_addr;
+
+
+	// get the target process' libc address and use it to find the
+	// addresses of the syscalls we want to use inside the
+	// target process
+	target_libc_addr = get_libc_addr(target);
+	target_libdl_addr = get_libdl_addr(target);
+	if (target_libdl_addr == 0)
+		pr_err_ns("Libdl does not loaded from target process");
+	target_malloc_addr = target_libc_addr + malloc_offset;
+	target_free_addr = target_libc_addr + free_offset;
+	target_dlopen_addr = target_libdl_addr + dlopen_offset;
+
+	memset(&oldregs, 0, sizeof(struct user_regs_struct));
+	memset(&regs, 0, sizeof(struct user_regs_struct));
+	ptrace_attach(target);
+	ptrace_getregs(target, &oldregs);
+	memcpy(&regs, &oldregs, sizeof(struct user_regs_struct));
+
+	/*
+	 * since ELF align each section size by paging size, many section
+	 * have padding at its tail. code be load shared object is
+	 * injected to here. and substract word size because ptrace read
+	 * word size at once.
+	 */
+	addr = get_inject_code_addr(target) - inject_content_size - sizeof(long);
```

Comment:
I don't understand why you subtracted a word size.  But it'd be better align it to a word size though.